akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -988,6 +988,35 @@ trait FlowOps[+Out, +Mat] {
    */
   def mapConcat[T](f: Out => IterableOnce[T]): Repr[T] = statefulMapConcat(() => f)
 
+  /**
+   * Transform each input element into a pair of an element of `T` and a state of `S`. The element will be emitted downstream
+   * and the state will be passed to the next step in the chain. The returned `T` MUST NOT be `null` as it is illegal
+   * as stream element - according to the Reactive Streams specification. For stateless variant see [[FlowOps.map]].
+   *
+   * The initial state is created using the given function `create` when the flow is materializing.
+   *
+   * The `onComplete` function is called only once when the upstream or downstream finished, You can do some clean-up here,
+   * and if the returned value is not empty, it will be emitted to the downstream if available, otherwise the value will be dropped.
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * '''Emits when''' the mapping function returns an element and downstream is ready to consume it
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   *
+   * @tparam S the type of the state
+   * @tparam T the type of the output elements
+   * @param create a function that creates the initial state
+   * @param f a function that transforms the upstream element and the state into a pair of next state and output element
+   * @param onComplete a function that transforms the ongoing state into an optional output element
+   */
+  def statefulMap[S, T](create: () => S, f: (S, Out) => (S, T), onComplete: S => Option[T]): Repr[T] =
+    via(new StatefulMap[S, Out, T](create, f, onComplete))
```

Comment:
should this be curried in Scala? like `bracket`

Reply:
I think latten `T = Option[Something]` seems better if there is no `async` lives bettwen, Will update ,thanks

Reply:
@johanandren @otto-dev I tried the uncurried version， the type inferrence is very very bad。

For Java I think that should be fixable with a phantom type?